RFD 100: Proxy gRPC transport

rosstimothy opened the pull request:
[Rendered](https://github.com/gravitational/teleport/blob/rfd/0100-proxy-ssh-grpc/rfd/0100-proxy-ssh-grpc.md)

rosstimothy committed:
RFD 100: Proxy gRPC transport

zmb3 commented on rfd/0100-proxy-ssh-grpc.md:
```diff
@@ -0,0 +1,197 @@
+
+The behavior of `tsh ssh` should remain the same regardless of the configured
+session recording mode. The time it takes to establish a session may be noticeably
+faster depending on proximity of the client and the Proxy.
```
Do you plan to add a fallback to the old double handshake method if the ALPN connection fails? Technically we don't support newer clients with older servers, but in practice it usually works.

espadolini replied:
+1 for this, we can't expect users to micromanage their `tsh` version, IMO.

espadolini commented on rfd/0100-proxy-ssh-grpc.md:
```diff
@@ -0,0 +1,197 @@
+to the gRPC server and all other requests to the SSH server.
+
+Note: a gRPC server is already exposed via the Proxy web address that users the ALPN protocol
+`teleport-proxy-grpc`. In order to not conflict the new ALPN protocol is required. Reusing the
+existing gRPC server is not an option since it is unauthenticated and has aggressive keep alive
+parameters.
```
Does `tsh` require connectivity to the web port? Only having to connect to the "ssh" proxy port would be another reason to have a separate server for this.

espadolini commented on rfd/0100-proxy-ssh-grpc.md:
```diff
@@ -0,0 +1,197 @@
+  string host = 1;
+  // The cluster the server is a member of
+  string cluster = 2;
+  // The OS login 
+  string login = 3;
```
What's this for?

espadolini replied:
In proxy recording mode, the OS login will be sent as part of the SSH handshake with the proxy, and outside of proxy recording mode the actual OS login that's going to be used in SSH is not exposed to the proxy, so the client could just lie about this.

espadolini replied:
IOW, when not in proxy recording mode, the client could literally just hardcode `-teleport-internal-join` here and the proxy would have to let it through.

espadolini commented on rfd/0100-proxy-ssh-grpc.md:
```diff
@@ -0,0 +1,197 @@
+
+// ClusterDetails contains details about the cluster configuration
+message ClusterDetails {
+  // If proxy recording mode is enabled
+  bool recording_proxy = 1;
```
If the proxy is malicious it can claim that proxy recording mode is enabled _anyway_; we could simply initialize the agent as soon as the proxy sends something to the agent, instead.

espadolini replied:
That's a discussion for another day, but if our model assumes that the proxy won't lie about `recording_proxy`, then we can also assume that the proxy simply won't ask the user's agent for signatures if it's not in proxy recording mode, so unless we need to know about proxy recording mode for some other reason, we can just drop this field IMO.

espadolini commented on rfd/0100-proxy-ssh-grpc.md:
```diff
@@ -0,0 +1,197 @@
+// Target indicates which server to connect to
+message Target {
+  // The server hostname/ip/uuid
+  string host = 1;
```
Is this a hostport, or are we disallowing direct connections to arbitrary hosts with this?

espadolini commented on rfd/0100-proxy-ssh-grpc.md:
```diff
@@ -0,0 +1,197 @@
+    // to the target host
+    Target dial_target = 1;
+    // Raw data to transmit
+    DataFrame frame = 2;
```
If this is `DialNode`, could we just have hardcoded `ssh_frame` and `agent_frame` messages instead? Other protocols could just use a different RPC in the future.

espadolini commented on rfd/0100-proxy-ssh-grpc.md:
```diff
@@ -0,0 +1,197 @@
+target exists and session control allows, the server will establish the
+connection and respond with a `ConnectionDetails` message.
+
+After the initial ceremony, either side of the connection may transmit `Frame`s
```
We could save a roundtrip if the server could send its version string together with the `ConnectionDetails`. Perhaps `DialNodeResponse` could just hold both messages, and it'd just be illegal for the server to send a non-null `ConnectionDetails` after its first message?

The same could be done for `DialNodeRequest`, but it would be useless for the SSH protocol.

espadolini replied:
With the SSH protocol it wouldn't be unexpected to have data (the `SSH-2.0-Teleport` line) available to send immediately after connecting; if we allowed the ConnectionDetails message together with a data frame we could send both in a single grpc stream message, which could save a TCP roundtrip and some overhead.

espadolini reviewed (commented):
I'd like to see either support or a path for future support for connection resumption, as @fspmarshall said.

Perhaps an optional connection token in `ConnectionDetails`, some counters in `DataFrame` to specify where the data is intended to start from, and some way to acknowledge the reception of data (with another counter in `DataFrame` maybe)?

zmb3 commented:
> I'm not sure that we should explicitly cover resumption in this RFD as that feels like feature creep

+1 - the goal we are pursuing with this work is to _reduce tsh connection latency_, not add a new feature. So long as the changes proposed here don't preclude us from adding connection resumption in the future we should consider that a non-goal here.

rosstimothy committed:
fix: clarify fallback and add port

fspmarshall commented on rfd/0100-proxy-ssh-grpc.md:
```diff
@@ -0,0 +1,199 @@
+  rpc GetClusterDetails(GetClusterDetailsRequest) returns (GetClusterDetailsResponse);
+  // DialNode establishes a connection to the target host
+  // 
+  // The bidirectional stream is a transport mechanism to send raw data between
+  // client and server. This allows other protocols(SSH) to be sent through the
+  // tunnel.
+  rpc DialNode(stream DialNodeRequest) returns (stream DialNodeResponse);
```
Nit: I feel that the `Node` terminology is too ambiguous.  I'd prefer to avoid using it in new code. IMO something like `DialSSH` would be better, or, if we wanted the option to add non-ssh protocols to the same stream, something like `DialAgent`/`DialService`.

fspmarshall replied:
Depends on wether we want to have a method per protocol, or extend this method to encompass other protocols.

My top picks would either be `DialSSH` or `DialService` depending on wether we're going specific or generic.

fspmarshall commented on rfd/0100-proxy-ssh-grpc.md:
```diff
@@ -0,0 +1,199 @@
+  PROTOCOL_UNSPECIFIED = 0;
+  // SSH
+  PROTOCOL_SSH = 1;
+  // SSH Agent
+  PROTOCOL_SSH_AGENT = 2;
```
Nit: Given how confusing it can be to talk about agent being forwarded to proxy, versus agent being forwarded to the ssh service, could we disambiguate this?  Maybe call it `PROTOCOL_SSH_AGENT_TO_PROXY`, or something like that?  I'd love for the "agent forwarded for the purpose of proxy impersonation of user" to be a bit more explicit.

espadolini replied:
I second Forrest's, specifying just the protocol name can be confusing.

fspmarshall replied:
I guess what bothers me, is that the two protocols have different destinations/purposes.  One is sideband information to be consumed by the proxy, and one is part of the `client <-> service` connection that just happens to be being forwarded by the proxy.

Looking at this, that's very non-obvious.  I see your point about keeping the enum agnostic so that we can reuse `DataFrame`.  Maybe there's somewhere else we could add some kind of destination discriminant?  Maybe we could use the same `DataFrame`, but extend the top-level enum to have `proxyFrame` and `serviceFrame` variants, instead of a single `frame`?

Idk, might be overthinking it.  I've just found myself disliking how opaque the naming in the proxy/reversetunnel system can be.

fspmarshall reviewed (approved):
Added a few minor suggestions.  Proposal looks good to me.

---

> I'd like to see either support or a path for future support for connection resumption

IIUC the `PROTOCOL_SSH` frames are not modified by the proxy when not in recording mode.  So long as that is true, I think we're good.  If we convert the `proxy <-> agent` side of the conn to GRPC in the future, then we can add a counter/id to all `PROTOCOL_SSH` frames, allowing client and agent to each send over a "last seen frame ID" during a resume operation.  As long as both sides hold onto a couple of their most recently sent frames, that should let us affect a basic resume in most cases I think (barring insanely high throughput).

rosstimothy committed:
- Refactor TransportService and make less generic
- Remove oneof from stream messages to pack multiple frames
- Add future considerations for session resumption

rosstimothy committed:
Merge branch 'master' into rfd/0100-proxy-ssh-grpc
